Treat non-JSON messages as fatal errors

alshakero opened the pull request:
Fixes: https://github.com/Palindrom/Palindrom/issues/188

Reminder: don't waste your time reviewing `dist` files. And don't worry about potential conflicts with https://github.com/Palindrom/Palindrom/pull/190, they're all in `dist` files.

warpech commented on src/palindrom.js:
```diff
@@ -331,8 +331,13 @@ const Palindrom = (() => {
+        }
+        catch (e) {
+          this.onFatalError(event.data, upgradeURL, 'WS');
+        }
```
Can you add a screenshot of how does it look with the server-side change from https://github.com/Starcounter/Home/issues/340?

@alemoi sends the stack trace in the text message and then sends the closing frame with the reason. Does it look decently readable in Palindrom or is it displayed as two verbose errors?

alemoi replied:
Yes, 2 frames, one message string with error and second Close frame that says: "The error stack trace is sent in a previous message."

alshakero replied:
I couldn't see the nice error message due to a `palindrom-polymer-client` limitation. 

Here is how it looks:
![image](https://user-images.githubusercontent.com/17054134/34102761-8b65a232-e3ea-11e7-9709-eb763133d951.png)

And that's because `palindrom-polymer-client` constructs the error message on its own and neglects `data` property from the event. 

See:
![image](https://user-images.githubusercontent.com/17054134/34102945-356ffc78-e3eb-11e7-80ef-86089d49b879.png)

alemoi commented:
I have produced a build: http://downloads.starcounter.com/download/Starcounter/Custom/2.4.0.4211

alshakero commented:
https://github.com/Palindrom/palindrom-polymer-client/issues/54 is fixed in https://github.com/Palindrom/palindrom-polymer-client/pull/55. Hers is the current output. Looks great!

![image](https://user-images.githubusercontent.com/17054134/34381788-fd14dd0e-eb0a-11e7-811b-a9bbffd001c9.png)

warpech commented:
I think that the first error row should start with some explanation: 

> `The Palindrom server responded with an error: ...`

WDYT?

alshakero commented:
Great idea.

After your comments on https://github.com/Palindrom/palindrom-polymer-client/pull/55 I think we should rethink this a little. 

The error messasge contruction in `palindrom-polymer-client` isn't clean at all IMO. I don't know why it has to construct its own error messages. I propose that we change that.

`onConnectionError` would have one parameter of type `PalindromError `:

```js
class PalindromError extends Error { 
	type: "validation|connection|generic",
	message: "an elaborate message",
	side: "server|local",
	[URL: "the perspective URL"], // if the error is network-related
}
```

If you agree, we can then fix `onGenericError` too and use the same type. I'd really like this because adding a new error type is terrible and hard to design.

warpech commented:
I like that!

Great to learn that ES6 allows extending `Error`! This was always [a pain](https://stackoverflow.com/questions/1382107/whats-a-good-way-to-extend-error-in-javascript).

Instead of `type: "validation|connection|generic",`, why don't we have: 

```txt
class PalindromError extends Error
class PalindromValidationError extends PalindromError
class PalindromConnectionError extends PalindromError
```

Instead of `side: "server|local"` please use the established vocabulary from Palindrom API and docs: `origin: "remote|local"`

Instead of `URL:` let's use `url:` because it is usually lowercase in modern Web APIs such as [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request/url).

alshakero commented:
Looking forward to Monday already. I love working on Palindrom.

alshakero commented:
@warpech @tomalec this is ready for review.

tomalec commented on src/palindrom.js:
```diff
@@ -331,8 +342,19 @@ const Palindrom = (() => {
+              event.data,
+              'Server',
+              this._ws.url,
+              'WS'
```
What if onReceive client side error is caught, should we still present it as server side error? I assume it's for non JSON responses, and JSON parsing errors, but I could expect onReceive callback errors could happen as well.

alshakero replied:
Very good point!

tomalec commented on src/palindrom.js:
```diff
@@ -341,13 +363,14 @@ const Palindrom = (() => {
+        const message = [
+          `WebSocket connection could not be made`,
+          ` readyState: ${this._ws.readyState}`
+        ].join('\n');
```
Why not use just multiline string?

alshakero replied:
Because the whitespace on the left would be included.

tomalec commented on src/palindrom.js:
```diff
@@ -358,18 +381,28 @@ const Palindrom = (() => {
+            ` reason: ${event.reason}`,
+            ` readyState: ${this._ws.readyState}`,
+            ` stateCode: ${event.code}`
+          ].join('\n');
```
Why not use just multiline string?

tomalec commented on src/palindrom.js:
```diff
@@ -358,18 +381,28 @@ const Palindrom = (() => {
+            ` reason: ${event.reason}`,
+            ` readyState: ${this._ws.readyState}`,
+            ` stateCode: ${event.code}`
+          ].join('\n');
```
Why not use just multiline string?

tomalec commented on src/palindrom.js:
```diff
@@ -396,10 +429,17 @@ const Palindrom = (() => {
+          ' Please reload the page.'
+          ` Previous session ID: ${this.remoteUrl}`
+          ` New session ID: ${remoteUrl}`
+        ].join('\n');
```
Why not use just multiline string?

tomalec commented on src/palindrom.js:
```diff
@@ -455,24 +495,28 @@ const Palindrom = (() => {
+            ` statusCode: ${statusCode}`,            
+            ` reason: ${reason}`,
+            ` url: ${res.url}`
+          ].join('\n');
```
Why not use just multiline string?

alshakero replied:
Because the whitespace on the left would be included.

alshakero commented:
@tomalec sorry for the trouble. I modified one more time. 

I dropped PalindromValidationError since there is a type already defined with all the needed details in [JSON-Patch](https://github.com/Starcounter-Jack/JSON-Patch/blob/master/src/helpers.ts#L148). And direction paramater was redundant since there are two seperate callbacks for each direction.

alshakero commented:
The minifier has an issue with arrays of ES6 template literals, I reported it https://github.com/webpack-contrib/babel-minify-webpack-plugin/issues/76

alshakero commented:
How errors look now:

### Restarting the app:

![connection error 404](https://user-images.githubusercontent.com/17054134/34825816-4c497216-f6d4-11e7-8101-cb5716b52ac7.jpg)

### Killing the server
![connection server restart](https://user-images.githubusercontent.com/17054134/34825818-4c7fc906-f6d4-11e7-860e-4c46b3331047.jpg)

### Sending an add patch:
![connection server error](https://user-images.githubusercontent.com/17054134/34825817-4c6468dc-f6d4-11e7-8cd7-e9398d63140e.jpg)

warpech commented on src/palindrom.js:
```diff
@@ -163,9 +167,16 @@ const Palindrom = (() => {
         return;
       }
-      scheduledError = setTimeout(() => {
+      var scheduledError = setTimeout(() => {
```
This looks wrong. `scheduledError` is declared on line 146.

warpech commented on src/palindrom.js:
```diff
@@ -163,9 +167,16 @@ const Palindrom = (() => {
+        onError(
+          new PalindromConnectionError(
+            "Timeout has passed and response hasn't arrived",
+            'Client',
```
Can this be a `const`?

alshakero replied:
I don't get it. Could you elaborate please?

tomalec replied:
like `const CLIENT = 'Client'; in line no 1, and just

warpech commented on src/palindrom.js:
```diff
@@ -163,9 +167,16 @@ const Palindrom = (() => {
+            "Timeout has passed and response hasn't arrived",
+            'Client',
+            this.remoteUrl,
+            'Unknown'
```
Can this be a `const`?

warpech commented on src/palindrom.js:
```diff
@@ -341,15 +363,17 @@ const Palindrom = (() => {
+        const message = [
+          'WebSocket connection could not be made',
+          'readyState: ' + this._ws.readyState
+        ].join('\n');
```
You can use:

alshakero replied:
Yes, do you thing it would look good without indentation? I don't mind either ways. But I lean to good looks here because perf is irrelevant, since the app is dead at this point.

warpech commented on src/palindrom.js:
```diff
@@ -777,8 +826,13 @@ const Palindrom = (() => {
-            this.onConnectionError(error);
+            this.onError(
+              new PalindromError(
+                `Palindrom: Error inside onStateReset callback: ${
```
Either one or none error texts should start with `Palindrom: `

warpech commented on docs/02-_Palindrom.md:
```diff
@@ -45,6 +45,7 @@ Attribute              | Type          | Default                | Description
 `onPatchSent`          | *Function*    |                        | Helper callback triggered each time a JSON-patch is sent, accepts two parameters: (*String* `data`, *String* `url`, *String*, `method`)
 `onSocketStateChanged` | *Function*    |                        | Helper callback triggered when socket state changes, accepts next parameters: (*int* `state`, *String* `url`, *String* `data`, *int* `code`, *String* `reason`)
+`onError`    | *Function*    |                        | Helper callback triggered when a generic error occurs. Called with PalindromError.
 `onConnectionError`    | *Function*    |                        | Helper callback triggered when socket connection closed, socket connection failed to establish, http requiest failed. Accepts next parameters: (*Object* `data`, *String* `url`, *String*, `method`). The data object contains the following properties: *String* `statusText` (HTTP response status code reason phrase or WebSocket error ti ...
```
The parameters of the `onConnectionError` and few other callbacks are now changed. Please update the README (and the Docs?)

alshakero commented:
I say number validation error message is much better this way 💃 
![image](https://user-images.githubusercontent.com/17054134/34870358-218342f8-f78a-11e7-8365-38f621b9fab5.png)
